Improve package signing plugin integration

danimtb opened the pull request:
Changelog: Feature: New `conan cache sign` and `conan cache verify` commands for signing and verifying packages.
Changelog Feature: Standardize package signing manifest and signatures metadata.
Docs: https://github.com/conan-io/docs/pull/4365

- Added new cache commands for sign and verify.
- Added text and JSON format output with package lists.
- Manage errors for the report.
- Added tools to create a signature manifest and a signatures metadata file in JSON format with files, signing method, and signing provider.

Here is an example plugin with OpenSSL using the features of this PR: https://github.com/conan-io/examples2/pull/207

danimtb committed:
[feature] Improve package signing plugin integration

danimtb committed:
wip

danimtb committed:
add tools and some fixes

danimtb committed:
add todo

danimtb committed:
add comment

danimtb committed:
minor fix

danimtb committed:
undo upload

danimtb committed:
review cache cmds

danimtb committed:
reduce function

danimtb committed:
fix

danimtb committed:
move logic to cache subapi

danimtb committed:
reduce code

danimtb committed:
simplify exception

danimtb committed:
simplify more

danimtb committed:
fix

danimtb committed:
clean

memsharded commented on conan/api/subapi/cache.py:
```diff
@@ -77,6 +80,69 @@ def check_integrity(self, package_list):
+                    pkg_signer.sign_pkg(rref, recipe_bundle.get("files", {}), rref_folder)
+                except Exception as e:
+                    recipe_bundle["error"] = e
+            for pref in packages:
```
Is it possible to have ``packages`` without ``recipe_bundle``? 
So it would become possible to sign the packages without signing the recipe?

danimtb replied:
I think both should be signed, as packages cannot be uploaded without their recipe, and the recipe is the proxy to packages, so it makes sense to have both signed

memsharded replied:
This is not what I meant, and it is connected to the above comment (that has been 👍, then resolved without a change nor response?):

> I think this should enforce a recipe_bundle to exist. If it doesn't exist it will skip it, doing nothing, giving a false sense of security "things are signed properly", but this is not true.

So the

> if recipe_bundle 

is still not fully clear to me

memsharded commented on conan/api/subapi/cache.py:
```diff
@@ -77,6 +80,69 @@ def check_integrity(self, package_list):
+        """Verify packages with the package signing plugin"""
+        cache = PkgCache(self._conan_api.cache_folder, self._api_helpers.global_conf)
+        pkg_signer = PkgSignaturesPlugin(cache, self._conan_api.home_folder)
+        if not pkg_signer.is_configured:
```
This doesn't seem correct, the only thing that is needed for this ``verify()`` is the verify function, but ``is_configured`` requires both functions to be define.

memsharded commented on conan/api/subapi/cache.py:
```diff
@@ -77,6 +80,69 @@ def check_integrity(self, package_list):
+        """Sign packages with the package signing plugin"""
+        cache = PkgCache(self._conan_api.cache_folder, self._api_helpers.global_conf)
+        pkg_signer = PkgSignaturesPlugin(cache, self._conan_api.home_folder)
+        if not pkg_signer.is_configured:
```
This doesn't seem correct, the only thing that is needed for this ``sign()`` is the verify function, but ``is_configured`` requires both functions to be define.

memsharded commented on conan/api/subapi/cache.py:
```diff
@@ -77,6 +80,69 @@ def check_integrity(self, package_list):
+                try:
+                    pkg_signer.sign_pkg(rref, recipe_bundle.get("files", {}), rref_folder)
+                except Exception as e:
+                    recipe_bundle["error"] = e
```
Maybe a ``sign_error`` would be more explicit?

memsharded replied:
I dont think adding Exceptions to package lists is good, as package lists needs to be serialzied/deserialized properly. I would restrict this to strings.

memsharded commented on conan/api/subapi/cache.py:
```diff
@@ -77,6 +80,69 @@ def check_integrity(self, package_list):
+                    try:
+                        pkg_signer.verify(pref, pref_folder, files)
+                    except Exception as e:
+                        pkg_bundle["error"] = e
```
Maybe a ``sign_verify_error``?

memsharded commented on conan/api/subapi/cache.py:
```diff
@@ -77,6 +80,69 @@ def check_integrity(self, package_list):
+
+        for rref, packages in package_list.items():
+            recipe_bundle = package_list.recipe_dict(rref)
+            if recipe_bundle:
```
I think this should enforce a ``recipe_bundle`` to exist. If it doesn't exist it will skip it, doing nothing, giving a false sense of security "things are signed properly", but this is not true.

memsharded replied:
Same as above

memsharded commented on conan/api/subapi/cache.py:
```diff
@@ -77,6 +80,69 @@ def check_integrity(self, package_list):
+
+        for rref, packages in package_list.items():
+            recipe_bundle = package_list.recipe_dict(rref)
+            if recipe_bundle:
```
I think this should enforce a ``recipe_bundle`` to exist. If it doesn't exist it will skip it, doing nothing, giving a false sense of security "things are signed properly", but this is not true.

memsharded commented on conan/cli/commands/list.py:
```diff
@@ -33,7 +33,7 @@ def print_serial(item, indent=None, color_index=None):
     if isinstance(item, dict):
         for k, v in item.items():
-            if isinstance(v, (str, int)):
+            if isinstance(v, (str, int, Exception)):
```
Doens't look great package lists need to be serialized/deserialized to json and text files.

memsharded commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -13,30 +14,38 @@ def __init__(self, cache, home_folder):
+        #  calling the plugins' sign function that provider and method fields are filled.
+        self._plugin_sign_function(ref, artifacts_folder=folder, signature_folder=metadata_sign)
+        for f in os.listdir(metadata_sign):
+            files[f"{METADATA}/sign/{f}"] = os.path.join(metadata_sign, f)
```
This doesn't seem a responsibility of the plugin? Collecting the files from the folder and adding them to the list of files to be uploaded.

memsharded commented on conan/tools/pkg_signing/plugin.py:
```diff
@@ -0,0 +1,64 @@
+SIGN_SUMMARY_FILENAME = "sign-summary.json"
+
+
+def get_summary_file_path(signature_folder):
```
So finally these tools are "recommendations" or available for users, but not enforced by Conan at all, users could be doing any signing the way they like, without summary paths, etc?

danimtb committed:
convert exception to str

danimtb committed:
error on empty pkglist

danimtb committed:
change plugin configured

czoido commented on conan/api/subapi/cache.py:
```diff
@@ -100,26 +103,28 @@ def sign(self, package_list):
                     pkg_signer.sign_pkg(rref, recipe_bundle.get("files", {}), rref_folder)
                 except Exception as e:
-                    recipe_bundle["error"] = e
+                    recipe_bundle["pkgsign_error"] = str(e)
```
Why not just `"error"` ? The string is what is going to provide the information about what type of error it is.

memsharded replied:
I recommended it based on the fact that you can have package lists with different errors due to different reasons, to give some clarity over if it is a signing error, a verify error, etc.

danimtb committed:
add signatures and manifest files

danimtb committed:
fix

danimtb committed:
fix enabled remotes

danimtb committed:
fix remotes

czoido commented on conan/cli/commands/cache.py:
```diff
@@ -5,12 +5,68 @@
 
+def _get_package_sign_error(pkg_list):
+    signs = []
+    for rref, packages in pkg_list.items():
+        recipe_bundle = pkg_list.recipe_dict(rref)
+        if recipe_bundle:
+            signs.append(recipe_bundle.get("pkgsign_error"))
+        for pref in packages:
+            pkg_bundle = pkg_list.package_dict(pref)
+            if pkg_bundle:
+                signs.append(pkg_bundle.get("pkgsign_error"))
+
+    if any((sign is not None) for sign in signs):
+        return ConanException("There were some errors in the package signing process. "
+                              "Please check the output.")
+    return None
```
Shouldn't this return when the first exception is found instead of iterating all the `pkg_list`?

czoido commented on conan/tools/pkg_signing/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    save(get_signatures_filepath(signature_folder), json.dumps(content, indent=2))
+
+
+def verify_files_checksums(signature_folder, files):
```
I'm not sure about this, I think this should iterate over the manifest, not the passed files, and if there's something missing fail? Similar to this? Could this happen?


```
def verify_files_checksums(signature_folder, files):
    expected_list = load_manifest(signature_folder).get("files", [])
    expected_files = {item["file"]: item["sha256"] for item in expected_list}

    for filename, expected_checksum in expected_files.items():
        if filename not in files:
            raise ConanException(f"File '{filename}' is listed in the manifest but missing from the package.")
        
        file_path = files[filename]
        actual_checksum = sha256sum(file_path)
        
        if actual_checksum != expected_checksum:
             raise ConanException(f"Checksum mismatch for file {filename}...")
```

danimtb replied:
yes, originally it was implemented that way but it was failing all the time because export_sources.tgz was not present during installation, as Conan skips downloading the sources by default if they are not needed 😅

danimtb replied:
I have added a specific test por this case and a note in the function

danimtb committed:
pkgsign error

memsharded commented on conan/api/model/list.py:
```diff
@@ -328,6 +328,15 @@ def items(self) -> Iterable[Tuple[RecipeReference, Dict[PkgReference, Dict]]]:
                         packages[pref] = prev_info
                 yield recipe, packages
 
+    def has_items(self) -> bool:
```
This is not necessary, users can check it with:

- Regular for-loop result checking in ``for rref, packages in package_list.items():`` with a flag or something
- Are we sure that it makes sense to raise an exception for an empty package list? What does ``conan upload`` do for an empty package list? And ``conan remove``?
- And even if users want to do explicit checking it is possible with ``not dict(mypkglist.items())``, not sure a new method is worth it, I'd prefer to keep the API as lean as possible.

danimtb replied:
With a pkglist like:
```json
{
    "Local Cache": {
        "pkg/0.1": {}
    }
}
```
- `conan remove`:
   ```
  WARN: Nothing to remove, package list do not contain recipe revisions
  Remove summary:
  Local Cache
   ```
- `conan upload`:
  ```
  Upload completed in 0s
  -------- Upload summary --------
  default
    pkg/0.1
  ```
  But a `conan search * -r default` shows that nothing was uploaded:
  ```
  Connecting to remote 'default' with user 'admin'
  Found 0 pkg/version recipes matching * in default
  default
    WARN: There are no matching recipe references
  ```

In case of a completely empty package list:
- `conan remove`: Same warning
- `conan upload`: `WARN: No packages were uploaded because the package list is empty.`

I will change this to a warning message and leave pkglist class untouched.

danimtb replied:
Changed this to an error when the package list is empty

memsharded commented on conan/cli/commands/cache.py:
```diff
@@ -5,12 +5,67 @@
+                                     "Please check the output.")
+    for rref, packages in pkg_list.items():
+        recipe_bundle = pkg_list.recipe_dict(rref)
+        if recipe_bundle:
```
Not sure why this ``if recipe_bundle``. 
The ``recipe_dict()`` method is called in a few places and no one checks the return for ``boolness`` of the return.

memsharded commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -12,34 +15,53 @@ def __init__(self, cache, home_folder):
+
     def sign(self, upload_data):
-        if self._plugin_sign_function is None:
+        if not self.is_sign_configured:
```
This is never happening, because ``if not pkg_signer.is_sign_configured:`` is checked in the public API before calling this internal signing.
Here, it should be an ``assert`` instead (or just remove it), as this is not intended as regular execution flow.

danimtb replied:
You are right regarding the cache sign/verify commands, but the sign/verify is also called in the upload/install commands, and this is why the check is needed at this point as well.

memsharded commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -12,34 +15,53 @@ def __init__(self, cache, home_folder):
 
     def verify(self, ref, folder, files):
-        if self._plugin_verify_function is None:
+        if not self.is_verify_configured:
```
Same as above, this should be an assert (if any, it can be just removed), not a no-op ``if``

memsharded commented on test/integration/test_pkg_signing.py:
```diff
@@ -54,3 +59,206 @@ def verify(ref, artifacts_folder, signature_folder, files):
+                                                    verify_files_checksums)
+
+        def sign(ref, artifacts_folder, signature_folder, **kwargs):
+            save(None, os.path.join(signature_folder, "pkgsign-manifest.json.sig"), "")
```
The usage of ``save()`` with ``None`` as a first argument is unsupported, or at the very least undocumented, and it could start failing at any time

danimtb replied:
Changed to internal `save()` as it is only for testing purposes

memsharded commented on test/integration/command/cache/test_cache_sign.py:
```diff
@@ -0,0 +1,247 @@
+    from conan.tools.files import save
+
+    def sign(ref, artifacts_folder, signature_folder, **kwargs):
+        save(None, os.path.join(signature_folder, "signature.sig"), "signed-content")
```
Usage of ``save(None`` is unsupported/undocumented, it could fail at any time

danimtb replied:
Changed to internal `save()` as it is only for testing purposes

memsharded commented on conan/tools/pkg_signing/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    """
+    Gets the path of the signatures file path
+    @param signature_folder: Signature folder path
+    @return: Path of the signatures file
```
Incorrect docstrings, should be ``:param ``

memsharded commented on conan/tools/pkg_signing/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    """
+    Loads the signatures file pkgsign-signatures.json from the signature folder
+    @param signature_folder: Signature folder path
+    @return: dict of {filename: signature_value}
```
Incorrect docstring

memsharded commented on conan/tools/pkg_signing/plugin.py:
```diff
@@ -0,0 +1,133 @@
+    """
+    Saves the content of signatures file in the signature folder
+    @param signature_folder: Signature folder path
+    @param signatures: dict of {filename: signature_value}
```
Docstring

memsharded commented on test/integration/test_pkg_signing.py:
```diff
@@ -54,3 +59,206 @@ def verify(ref, artifacts_folder, signature_folder, files):
+
+        def verify(ref, artifacts_folder, signature_folder, files, **kwargs):
+            print(f"Manifest content:\n {load_manifest(signature_folder)}")
+            verify_files_checksums(signature_folder, files)
```
I don't fully understand why the checksums validation doesn't happen internally in the same way it is created internally.

danimtb replied:
Added checksums validation check before calling the user's plugin

danimtb committed:
remove recipe_bundle check

danimtb committed:
fix docstrings

danimtb committed:
warn on empty pkglist

danimtb committed:
fix

danimtb committed:
use interal save() only in tests

danimtb committed:
verify manifest checksums before calling the plugin

danimtb committed:
add test with conan_sources.tgz manifest validation

danimtb committed:
Merge branch 'develop2' into danimtb/improve-pkg-sign2

danimtb committed:
fix import

czoido commented on conan/tools/pkg_signing/plugin.py:
```diff
@@ -0,0 +1,139 @@
+    :param signatures: dict of {filename: signature_value}
+    """
+    assert isinstance(signatures, list),\
+        "The signature plugin function must return a list of signatures values"
```
I don't think this is the right place to check this, I think it's more clear if we do this check outside this helper and also I would raise a ConanException instead of an assert

danimtb replied:
Moved to the plugin manager class

danimtb committed:
move signatures check

danimtb committed:
remove test not neededç

danimtb committed:
remove pkgsign tools

danimtb committed:
raise on empty pkglist

danimtb committed:
do not brake old behavior

danimtb committed:
change signatures file format

danimtb committed:
add deprecation warning for pkg sign in upload command

danimtb commented:
Implemented in the latest commits:
- Removed package signing "tools".
- Keep support for the old package sign plugin interface (`sign()` function doesn't return any data).
- Error when operating over an empty package list.
- Change `pkgsign-signatures.json` format: Now users should return the manifest file in the signature entry if they want to include it in the package.
- Added a deprecation warning in `conan upload` (only when users have the plugin configured), to note that the implicit signing of packages will be removed in a future Conan version.

czoido commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -1,9 +1,97 @@
+    if not os.path.isfile(os.path.join(signature_folder, PKGSIGN_MANIFEST)):
+        ConanOutput().warning(f"[Package sign] Manifest file does not exist in signature folder "
+                              f"{os.path.join(signature_folder, PKGSIGN_MANIFEST)}. "
+                              f"Skipping checksum verification.")
```
I think this message should be more clear about the fact that we have updated the way of how the plugins work and that you should update.
Also, I think the message would be more correct if it's like:

`Manifest file '{PKGSIGN_MANIFEST}' does not exist in signature folder '{signature_folder}'.` ?

czoido replied:
Maybe add `deprecated` tag here as well?

danimtb replied:
done

czoido commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -12,35 +100,62 @@ def __init__(self, cache, home_folder):
+                    files[f"{METADATA}/sign/{f}"] = os.path.join(metadata_sign, f)
+            ConanOutput().warning("[Package sign] The signature plugin sign() function must return "
+                                  "a list of signature dicts. See the documentation at "
+                                  "https://docs.conan.io/2/reference/extensions/package_signing.html")
```
Something like `"to update your plugin check the docs at..."`?
Also maybe adding `deprecated` tag?

AbrilRBS replied:
+1 to at least add a deprecated tag

danimtb replied:
done

czoido commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -1,9 +1,97 @@
+    expected_files = {item["file"]: item["sha256"] for item in expected_list}
+
+    # This is checking that the files of the package exist in the manifest instead of the opposite
+    # because some files might be missing such as conan_sources.tgz
```
I think we should add a warning in case there's some file missing?

danimtb replied:
done

czoido commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -12,35 +100,62 @@ def __init__(self, cache, home_folder):
+    def is_verify_configured(self):
+        return self._plugin_verify_function is not None
+
+    def sign_pkg(self, ref, files, folder):
```
Maybe we should clean existing signatures in case they exist so if we change signing method in between two different signs there are not unexpected files there that could be uploaded by mistake?

danimtb replied:
I thought about this at an early stage, but I think we should not do this by default because there are cases when you don't want this to be cleaned, for example:

- packages signed by another provider (maybe with a different sign method): You don't want to remove their signature by mistake!
- packages that are already signed (maybe the user wants to avoid resigning them for security).

There might be other cases, but I believe we should leave this to the users plugin's implementation,

Maybe a new command to clean the signatures might be needed (might be a very rare case or while developing and testing the plugin). Or even have `conan cache sign --force` to force the clean of the folder and resign the package

AbrilRBS commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -1,9 +1,97 @@
+    """
+    files_list = []
+
+    for fname in os.listdir(artifacts_folder):
```
```suggestion
    for fname in sorted(os.listdir(artifacts_folder)):
```

maybe cleaner?

AbrilRBS replied:
(and remove the sort below)

danimtb replied:
thanks! Done

AbrilRBS reviewed (commented):
super minor nitpick

danimtb committed:
update warnings

danimtb committed:
Merge remote-tracking branch 'upstream/develop2' into danimtb/improve-pkg-sign2

czoido commented on conan/cli/commands/cache.py:
```diff
@@ -158,6 +212,81 @@ def cache_check_integrity(conan_api: ConanAPI, parser, subparser, *args):
+        package_list = conan_api.list.select(ref_pattern, package_query=args.package_query)
+
+    if not dict(package_list.items()):
+        raise ConanException("No packages to process in the pkglist provided")
```
```suggestion
        raise ConanException("No packages to process in the package list provided")
```

czoido replied:
I think we use this term when we refer to those lists.

czoido commented on conan/cli/commands/cache.py:
```diff
@@ -158,6 +212,81 @@ def cache_check_integrity(conan_api: ConanAPI, parser, subparser, *args):
+        package_list = conan_api.list.select(ref_pattern, package_query=args.package_query)
+
+    if not dict(package_list.items()):
+        raise ConanException("No packages to process in the pkglist provided")
```
```suggestion
        raise ConanException("No packages to process in the package list provided")
```

czoido commented:
I can see that the JSON output of the `sign` and the `verify` is a bit asymmetric. The `sign` has the information about the signed files but the `verify` does not have information about the verified files, shouldn't we add that information?

danimtb committed:
remove sign on upload

danimtb committed:
fix json output asymmetry

danimtb committed:
Update conan/cli/commands/cache.py

Co-authored-by: Carlos Zoido <mrgalleta@gmail.com>

danimtb committed:
Update conan/cli/commands/cache.py

Co-authored-by: Carlos Zoido <mrgalleta@gmail.com>

danimtb committed:
Merge branch 'develop2' into danimtb/improve-pkg-sign2

danimtb committed:
Merge branch 'danimtb/improve-pkg-sign2' of github.com:danimtb/conan into danimtb/improve-pkg-sign2

danimtb committed:
update tests

AbrilRBS reviewed (approved):
I really like this new approach, good job!

memsharded commented on conan/internal/rest/pkg_sign.py:
```diff
@@ -1,9 +1,105 @@
+    expected_list = json.loads(manifest_content).get("files", [])
+    expected_files = {item["file"]: item["sha256"] for item in expected_list}
+
+    # Warn in case there are missing files (like conan_sources.tgz not being downloaded)
```
I don't see a test covering this functionality.

I am not even sure this is required, this will basically warn for every downloaded recipe, as ``conan_sources.tgz`` is never downloaded by default with the recipe, so when the verify() runs, it will always report about this?

danimtb replied:
ok, i think it is better to remove the warning, as the files will be checked anyway whenever something is downloaded

danimtb committed:
remove warning
